### PR TITLE
[CALCITE-7042] Eliminate nested TRIM calls, exploiting the fact that TRIM is idempotent

### PR DESCRIPTION
**Pull request by xuzifu666**
Jira: https://issues.apache.org/jira/browse/CALCITE-7042

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 1]
```diff
@@ -2364,6 +2366,37 @@ private RexNode simplifyCeilFloor(RexCall e) {
...
+   *
+   * <li>{@code trim(BOTH ' ' from trim(BOTH ' ' from ' aa '))}
+   * returns {@code trim(BOTH ' ' from ' aa ')}
+   *
+   * <li>{@code trim(LEADING 'a' from trim(BOTH ' ' from ' aa '))} does not change
+   *
+   * </ul>
+   */
+  private RexNode simplifyTrim(RexCall e) {
+    if (e.getOperands().size() != 3) {
```
Comment (silundong) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2384]: `SqlTrimFunction` seems to check that its operands must be three. When we will get here?

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 2]
```diff
@@ -2364,6 +2366,37 @@ private RexNode simplifyCeilFloor(RexCall e) {
...
+      return e;
+    }
+
+    if (e.getOperands().get(2) instanceof RexCall) {
```
Comment (silundong) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2388]: It seems that only idempotence is handled here, without simplifying the second and third operands.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java (hunk 1 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2384]: Yes, maybe we not need the check here because numbers of arg must be 3.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java (hunk 2 above)
Comment (silundong) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2388]: I think the simplification of idempotency alone is not sufficient. It should be combined with general simplification. For example, if the third operand is not trim but another `RexCall`, there is a lack of simplification for it.

Comment (silundong) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2388]: For example :
`trim(both case when ceil(...) > ... then '...' else '...' end from case when floor(...) < ... then ... else ... end)`
Perhaps this usage is rare, I just considered it as a possibility.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 3]
```diff
@@ -2364,6 +2366,38 @@ private RexNode simplifyCeilFloor(RexCall e) {
...
+      RexCall childNode = (RexCall) e.getOperands().get(2);
+      // only strings with the same trim method and deduplication will be eliminated.
+      if (childNode.getKind() == SqlKind.TRIM
+          && e.operands.get(0).equals(childNode.operands.get(0))
+          && e.operands.get(1).equals(childNode.operands.get(1))) {
+        return simplifyTrim((RexCall) e.getOperands().get(2));
```
Comment (xiedeyantu) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2394]: Maybe the third parameter needs to recursively call simplify, or determine whether the third parameter is a literal?

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 4]
```diff
@@ -2364,6 +2366,38 @@ private RexNode simplifyCeilFloor(RexCall e) {
...
+    if (e.getOperands().size() != 3) {
+      return e;
+    }
+
+    if (e.getOperands().get(2) instanceof RexCall) {
+      RexCall childNode = (RexCall) e.getOperands().get(2);
+      // only strings with the same trim method and deduplication will be eliminated.
+      if (childNode.getKind() == SqlKind.TRIM
+          && e.operands.get(0).equals(childNode.operands.get(0))
+          && e.operands.get(1).equals(childNode.operands.get(1))) {
```
Comment (xiedeyantu) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2393]: Maybe the childNode do not have two operands?

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java (hunk 3 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Here need only handle when the third operand is trim RexCall and recursively call by simplifyTrim function.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 5]
```diff
@@ -2804,6 +2804,69 @@ private SqlDialect nonOrdinalDialect() {
...
+        .withPresto().ok(exptectedPresto5);
+
+    final String query6 = "SELECT TRIM(LEADING 'a' from TRIM(BOTH 'a' from \"brand_name\"))\n"
+        + "from \"foodmart\".\"product\"";
+    final String exptectedPresto6 = "SELECT LTRIM(TRIM(\"brand_name\", 'a'), 'a')\n"
+        + "FROM \"foodmart\".\"product\"";
+    sql(query6)
+        .withPresto().ok(exptectedPresto6);
+
+    final String query7 = "SELECT TRIM(TRIM(BOTH 'a' from \"brand_name\"))\n"
```
[the remark this answers is not in the document]
Comment (xuzifu666) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: @mihaibudiu I had added more test about null illegal test cases, which would return NULL as expected. Thanks.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 6]
```diff
@@ -2364,6 +2366,38 @@ private RexNode simplifyCeilFloor(RexCall e) {
...
+  private RexNode simplifyTrim(RexCall e) {
+    if (e.getOperands().size() != 3) {
+      return e;
+    }
+
+    if (e.getOperands().get(2) instanceof RexCall) {
+      RexCall childNode = (RexCall) e.getOperands().get(2);
+      // only strings with the same trim method and deduplication will be eliminated.
+      if (childNode.getKind() == SqlKind.TRIM
+          && simplify(e.operands.get(0)).equals(simplify(childNode.operands.get(0)))
```
[the remark this answers is not in the document]
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2392]: It has been specified as <BOTH><LEADING><TAILING> in JAVACC file.
If this place is changed to an expression or column, a syntax error will be reported directly, so it may not be possible to add a test like this, but I understand that there is no problem with the simplified call here（because it must be a string literal.  I wonder if you agree with my point of view? @mihaibudiu

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2392]: ok, change my claim above for operands.get(1).

[the remark this answers is not in the document]
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java:2392]: @mihaibudiu OK，Had added test case for operand(1) at test end.

**Review by mihaibudiu — APPROVED** [review]

**Review by julianhyde — CHANGES_REQUESTED** [review]
This is a feature. The summary should reflect thst. “Eliminate nested TRIM calls, exploiting the fact that TRIM id idempotent”. 

A lot of test code has been added to RelToSqlConverterTest but this issue is not about conversion to a SQL dialect. Put it somewhere better. 


The sql queries in planner.iq need comments. Say which ones you expect the plan to fold calls, and which ones you expect it to not fold, and why. Otherwise your work will be lost when someone updates the logs. 

Is there a general mechanism for declaring thst functions are ifempotent? Can you give examples of other idempotent functions? I would like you to start thinking about that in the jira case even though there is no code yet.

File: core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java [hunk 7]
```diff
@@ -360,6 +361,97 @@ public class RexImplicationCheckerTest {
         hasToString("2014"));
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-7042">[CALCITE-7042]
+   * Eliminate nested TRIM calls, exploiting the fact that TRIM is idempotent</a>. */
+  @Test void testSimplifyIdempotentFunctions() {
```
Comment (xiedeyantu) [inline core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java:367]: I think we should pass a test case like the following, even if the third parameter in trim is not a trim expression, we should continue to recursively simplify. The floor I used may not be very appropriate, but I just want to express this idea.
```
  @Test void testSimplifyIdempotentFunctions() {
    final Fixture f = new Fixture();
    RexLiteral trimBoth = f.rexBuilder.makeLiteral("BOTH");
    RexLiteral trimed = f.rexBuilder.makeLiteral("a");

    final RexNode literalTs =
        f.timestampLiteral(new TimestampString("2010-10-10 00:00:00"));
    final RexNode innerFloorCall =
        f.rexBuilder.makeCall(SqlStdOperatorTable.FLOOR, literalTs,
            f.rexBuilder.makeFlag(TimeUnitRange.YEAR));
    final RexNode outerFloorCall =
        f.rexBuilder.makeCall(SqlStdOperatorTable.FLOOR, innerFloorCall,
            f.rexBuilder.makeFlag(TimeUnitRange.YEAR));
    RexCall floorSimplifiedExpr =
        (RexCall) f.simplify.simplifyPreservingType(outerFloorCall,
            RexUnknownAs.UNKNOWN, true);

    assertThat(floorSimplifiedExpr,
        hasToString("FLOOR(2010-10-10 00:00:00, FLAG(YEAR))"));

    RelDataType varcharType =
        f.typeFactory.createSqlType(SqlTypeName.VARCHAR);
    RexCall floorSimplifiedExpr2 =
        (RexCall) f.simplify.simplifyPreservingType(f.rexBuilder.makeCast(varcharType, outerFloorCall),
            RexUnknownAs.UNKNOWN, true);

    assertThat(floorSimplifiedExpr2,
        hasToString("CAST(FLOOR(2010-10-10 00:00:00, FLAG(YEAR))):VARCHAR NOT NULL"));

    RexCall innerTrimCall =
        (RexCall) f.rexBuilder.makeCall(SqlStdOperatorTable.TRIM, trimBoth, trimed,
            f.rexBuilder.makeCast(varcharType, outerFloorCall));
    RexCall outerTrimCall =
        (RexCall) f.rexBuilder.makeCall(SqlStdOperatorTable.TRIM, trimBoth, trimed, innerTrimCall);
    RexCall trimSimplifiedCall =
        (RexCall) f.simplify.simplifyPreservingType(outerTrimCall,
            RexUnknownAs.UNKNOWN, true);

    // We will get this Currently
    // assertThat(trimSimplifiedCall,
        hasToString("TRIM('BOTH', 'a', CAST(FLOOR(FLOOR(2010-10-10 00:00:00, "
                + "FLAG(YEAR)), FLAG(YEAR))):VARCHAR NOT NULL)"));

    // I think We hope to get like this
    assertThat(trimSimplifiedCall,
        hasToString("TRIM('BOTH', 'a', CAST(FLOOR(2010-10-10 00:00:00, "
            + "FLAG(YEAR))):VARCHAR NOT NULL)"));
  }
```

Comment (xuzifu666) [reply core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java:367]: Thanks for your idea and case, it is indeed better, here use floor to test is OK IMO and I had changed with test case. @xiedeyantu

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 8]
```diff
@@ -2364,6 +2366,43 @@ private RexNode simplifyCeilFloor(RexCall e) {
...
+          && simplify(e.operands.get(1)).equals(simplify(childNode.operands.get(1)))) {
+        return simplifyTrim(childNode);
+      }
+    }
+
+    ArrayList<RexNode> rexNodes = new ArrayList<>();
+    for (RexNode node : e.operands) {
+      rexNodes.add(simplify(node));
```
Comment (xiedeyantu) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: This may simplify the same RexNode twice. Since it will be done once in the above if, we can just do it once.

Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: OK, had optimzed it.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 9]
```diff
@@ -2364,6 +2366,45 @@ private RexNode simplifyCeilFloor(RexCall e) {
...
+   * <li>{@code trim(LEADING 'a' from trim(BOTH ' ' from ' aa '))} does not change
+   *
+   * </ul>
+   */
+  private RexNode simplifyTrim(RexCall e) {
+    if (e.getOperands().size() != 3) {
+      return e;
+    }
+
+    RexNode trimWay = simplify(e.operands.get(0));
```
Comment (xiedeyantu) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: trimWay -> trimType ?

**Review by xiedeyantu — APPROVED** [review]
There is no problem overall. You can adjust the variable naming and optimize the comments.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java (hunk 9 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Done, thanks.

**Comment by xiedeyantu** [comment]
Let's wait and see if @mihaibudiu and @julianhyde have any more comments.

**Comment by xiedeyantu** [comment]
You can squash this pr. we can wait for another 48 hours. If there are no other comments, I will merge this PR.

**Comment by xuzifu666** [comment]
> You can squash this pr. we can wait for another 48 hours. If there are no other comments, I will merge this PR.

Done，thanks~ @xiedeyantu

File: core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java [hunk 10]
```diff
@@ -360,6 +361,135 @@ public class RexImplicationCheckerTest {
...
+        (RexCall) f.rexBuilder.makeCall(SqlStdOperatorTable.TRIM, trimBoth, trimed, trimNull);
+    outerNullTrimCall =
+        (RexCall) f.rexBuilder.makeCall(SqlStdOperatorTable.TRIM, trimLeft, trimed, innerTrimCall);
+    trimSimplifiedNullCall =
+        f.simplify.simplifyPreservingType(outerNullTrimCall,
+            RexUnknownAs.UNKNOWN, true);
+    // after simplify trimSimplifiedCall is equal to null
+    assertThat(trimSimplifiedNullCall, hasToString("null:VARCHAR(2)"));
+
+    final RexNode literalTs =
```
Comment (xiedeyantu) [inline core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java]:
```suggestion
    // simplifyTrim supports recursive simplification
    RelDataType varcharType =
        f.typeFactory.createSqlType(SqlTypeName.VARCHAR);
    // castCall is CAST(CAST(1111):VARCHAR NOT NULL):VARCHAR NOT NULL
    final RexNode castCall =
        f.cast(varcharType,
            f.cast(varcharType, f.literal(1111)));

    innerTrimCall =
        (RexCall) f.rexBuilder.makeCall(SqlStdOperatorTable.TRIM, trimBoth, trimed, castCall);
    // outerTrimCall is
    // TRIM('BOTH', 'a', TRIM('BOTH', 'a', CAST(CAST(1111):VARCHAR NOT NULL):VARCHAR NOT NULL))
    outerTrimCall =
        (RexCall) f.rexBuilder.makeCall(SqlStdOperatorTable.TRIM, trimBoth, trimed, innerTrimCall);
    trimSimplifiedCall =
        (RexCall) f.simplify.simplifyPreservingType(outerTrimCall,
            RexUnknownAs.UNKNOWN, true);

    assertThat(trimSimplifiedCall,
        hasToString("TRIM('BOTH', 'a', '1111':VARCHAR)"));
```

File: core/src/test/resources/sql/planner.iq [hunk 11]
```diff
@@ -147,6 +147,81 @@ EnumerableCalc(expr#0..2=[{inputs}], $f0=[$t1], $f1=[$t2])
 !plan
 !set planner-rules original
 
+# [CALCITE-7042] Eliminate nested TRIM calls, exploiting the fact that TRIM is idempotent
```
Comment (xiedeyantu) [inline core/src/test/resources/sql/planner.iq]: All the cases should be move to functions.iq. You can squashed to one sql, like `select trim(trim(' aa ')), trim(trim(both 'c' from ' aa ')), ...;`.

File: core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java (hunk 10 above)
Comment (xiedeyantu) [reply core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java]: Sorry, all rows from this row onwards need to be replaced. I only selected one row.

**Comment by mihaibudiu** [comment]
Iq files check the final result but won’t check that the optimization is applied

**Comment by xuzifu666** [comment]
> Iq files check the final result but won’t check that the optimization is applied

testSimplifyIdempotentFunctions can  check the final result, maybe here not need iq test, I remove them. @mihaibudiu

File: core/src/test/resources/sql/planner.iq (hunk 11 above)
Comment (xuzifu666) [reply core/src/test/resources/sql/planner.iq]: iq test is unnecessary here, so I remove them.

File: core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java (hunk 10 above)
Comment (xuzifu666) [reply core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java]: OK,  had changed the case.

**Comment by xuzifu666** [comment]
@mihaibudiu I had changed the code, could you help take a look again?

File: core/src/main/java/org/apache/calcite/rex/RexInterpreter.java [hunk 12]
```diff
@@ -339,6 +341,25 @@ private static Comparable coalesce(List<Comparable> values) {
...
+    }
+    NlsString trimType = (NlsString) values.get(0);
+    NlsString trimed = (NlsString) values.get(1);
+    NlsString trimString = (NlsString) values.get(2);
+    switch (trimType.getValue()) {
+    case "BOTH":
+      return trimString.trim(trimed.getValue());
+    case "LEADING":
+      return trimString.ltrim(trimed.getValue());
+    case "TAILING":
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/rex/RexInterpreter.java]: this seem like a typo.
which suggests that this case is not tested.
Why is this function needed? I don't see how it's related to the issue at hand.

File: core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java [hunk 13]
```diff
@@ -360,6 +361,116 @@ public class RexImplicationCheckerTest {
         hasToString("2014"));
   }
 
+  /** Test case for
```
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java:364]: why are these functions in RexImplicationCheckerTest?

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/rex/RexInterpreter.java (hunk 12 above)
Comment (xuzifu666) [reply core/src/main/java/org/apache/calcite/rex/RexInterpreter.java]: thanks，had fixed it and added the test cover it from line 373 in RexImplicationCheckerTest##testSimplifyIdempotentFunctions

File: core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java (hunk 13 above)
Comment (xuzifu666) [reply core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java:364]: 1. Refer to Julian's suggestion, remove the test from RelToSqlConverterTest, and then refer to other Simplify tests such as testSimplifyCeilFloor, which add the test to RexImplicationCheckerTest, which seems to be no problem;
2. As for why ```RexInterpreter##trim``` is added, it is because ```rexBuilder.makeCall``` is needed to build TRIM, otherwise calling ```f.rexBuilder.makeCall(SqlStdOperatorTable.TRIM, f.rexBuilder.makeLiteral(trim), trimed, trimString);``` will error  out:
```
unbound: TRIM('BOTH', 'a', 'bb')
java.lang.IllegalArgumentException: unbound: TRIM('BOTH', 'a', 'bb')
	at org.apache.calcite.rex.RexInterpreter.unbound(RexInterpreter.java:99)
	at org.apache.calcite.rex.RexInterpreter.visitCall(RexInterpreter.java:244)
	at org.apache.calcite.rex.RexInterpreter.visitCall(RexInterpreter.java:58)
	at org.apache.calcite.rex.RexCall.accept(RexCall.java:208)
	at org.apache.calcite.rex.RexVisitor.visitList(RexVisitor.java:67)
```
@mihaibudiu

Comment (mihaibudiu) [reply core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java:364]: It's probably not the right class for the tests, but indeed, the other tests don't belong here either. So perhaps you can add them here.

Here is the JavaDoc of RexInterpreter:
```
 * Evaluates {@link RexNode} expressions.
 *
 * <p>Caveats:
 * <ul>
 *   <li>It uses interpretation, so it is not very efficient.
 *   <li>It is intended for testing, so does not cover very many functions and
 *   operators. (Feel free to contribute more!)
 *   <li>It is not well tested.
 * </ul>
```

So this class is not really designed for optimizing programs, rather for testing programs. But perhaps your use of it is just for testing.

Comment (xuzifu666) [reply core/src/test/java/org/apache/calcite/test/RexImplicationCheckerTest.java:364]: Yes, you are right, it is just for test here, does not affect the core logic.

**Review by mihaibudiu — COMMENTED** [review]
You need to squash the commits for this to be merged

**Comment by xuzifu666** [comment]
> You need to squash the commits for this to be merged

Had squahsed commits, Thanks @mihaibudiu
